streams/src/test/java/org/apache/kafka/streams/tests/EosTestDriver.java

```diff
@@ -250,56 +277,95 @@ private static void addRecord(final ConsumerRecord<byte[], byte[]> record,
+        final boolean validTopic = "data".equals(topic) || "echo".equals(topic) || "min".equals(topic) || "sum".equals(topic);
+
+        if (withRepartitioning) {
+            return validTopic || "repartition".equals(topic) || "max".equals(topic) || "cnt".equals(topic);
+        }
+
+        return validTopic;
+    }
 
+    private static boolean receivedAllRecords(final Map<String, Map<TopicPartition, List<ConsumerRecord<byte[], byte[]>>>> receivedRecords,
+                                              final Map<TopicPartition, Long> committedOffsets,
+                                              final boolean withRepartitioning) {
+        if (receivedRecords == null) {
             return false;
         }
 
-        for (final Map.Entry<TopicPartition, List<ConsumerRecord<byte[], byte[]>>> partitionRecords : receivedRecords.entrySet()) {
-            final TopicPartition tp = partitionRecords.getKey();
-            final int numberOfReceivedRecords = partitionRecords.getValue().size();
-            final Long committed = committedOffsets.get(new TopicPartition("data", tp.partition()));
-            if (committed != null) {
-                if (numberOfReceivedRecords < committed) {
+        // data -> data, echo, min, sum (repartition)
+        // repartition -> max, cnt
+
+        final Map<String, List<String>> inputToOutputTopics = new HashMap<>();
+        {
+            final List<String> dataOutputTopics = new ArrayList<>();
+            inputToOutputTopics.put("data", dataOutputTopics);
+
+            dataOutputTopics.add("data");
+            dataOutputTopics.add("echo");
+            dataOutputTopics.add("min");
+            dataOutputTopics.add("sum");
+
+            if (withRepartitioning) {
+                dataOutputTopics.add("repartition");
+
+                final List<String> repartitionOutputTopics = new ArrayList<>();
+                inputToOutputTopics.put("repartition", repartitionOutputTopics);
+                repartitionOutputTopics.add("max");
+                repartitionOutputTopics.add("cnt");
+            }
+        }
+
+        for (final Map.Entry<TopicPartition, Long> committed : committedOffsets.entrySet()) {
+            final TopicPartition inputTp = committed.getKey();
+            final String topic = inputTp.topic();
+            final int partition = inputTp.partition();
+            final long offset = committed.getValue();
+
+            for (final String outputTopic : inputToOutputTopics.get(topic)) {
+                final Map<TopicPartition, List<ConsumerRecord<byte[], byte[]>>> received = receivedRecords.get(outputTopic);
+                if (received == null) {
+                    return false;
+                }
+
+                final TopicPartition outputTp = new TopicPartition(outputTopic, partition);
+                final List<ConsumerRecord<byte[], byte[]>> records = received.get(outputTp);
+                if (records == null || records.size() < offset) {
```

Comment:
This check `records.size() < offset` seems a bit sketchy to me. Basically we are assuming that the input topic `data`'s starting offset is always 0, and there is no "holes" in the topic partitions so that the `offset` indicates the number of records we can read from the input topic.

Maybe a more robust way to do that would be 

1) read the input topics `data` and optionally `repartition` based on `withRepartitioning`, stop when the current record's offset is equal to or larger than the committed offset, and remember the number of records; 
2) read the output topics (again optionally `repartition`) from the beginning to the end (use `seekTo`), and check that the number of records are the same as the number of records read from the input.

Then we do not need to truncate, and also in verification we do not need to check list size again since they are already checked here.